perf(ivy): chain multiple attribute instructions

crisbeto opened the pull request:
Similarly to what we did in #31078, these changes implement chaining for the `attribute` instruction.

This PR resolves FW-1390.

kara commented on packages/core/test/acceptance/attributes_spec.ts:
```diff
@@ -73,6 +73,104 @@ describe('attribute binding', () => {
+    const a = fixture.debugElement.query(By.css('a')).nativeElement;
+    // NOTE: different browsers will add `//` into the URI.
+    expect(a.getAttribute('href')).toBe('https://angular.io/robots.txt');
+    expect(a.getAttribute('id')).toBe('my-link');
```
Nit: Since this is a property binding, perhaps it makes sense to test the property here too? Same for title.

kara commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1083,23 +1092,17 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-      nodeIndex: number, span: ParseSourceSpan|null, reference: o.ExternalReference,
-      callsOrFn?: o.Expression[][]|(() => o.Expression[][])) {
+      nodeIndex: number, reference: o.ExternalReference,
+      propertyBindings: ChainableBindingInstruction[]) {
```
Nit: Since this can receive attribute bindings, perhaps we should generalize the name?

```suggestion
      bindings: ChainableBindingInstruction[]) {
```

kara reviewed (approved):
LGTM, with some nits. And IIRC, host bindings part is being implemented in next PR, right?

kara commented:
[VE presubmit](http://test/OCL:254763918:BASE:254763946:1561390679500:b8c86f7f)

[Ivy presubmit](http://test/OCL:254763918:BASE:254764081:1561390735727:4e4b928)

crisbeto commented:
Reworked based on the feedback.

kara commented:
@crisbeto Seeing some G3 failures with Ivy, need to investigate

kara commented on packages/core/src/render3/instructions/attribute.ts:
```diff
@@ -34,6 +34,7 @@ export function ɵɵattribute(
-    return elementAttributeInternal(index, name, bound, lView, sanitizer, namespace);
+    elementAttributeInternal(index, name, bound, lView, sanitizer, namespace);
   }
+  return ɵɵattribute;
```
Due to a Tsickle bug, you will want to use the `TsickleIssue1009` type as the return type:

```ts
export function ɵɵattribute(
    name: string, value: any, sanitizer?: SanitizerFn | null, namespace?: string): TsickleIssue1009 {
```

This is what's causing the G3 failures

crisbeto replied:
Done.

crisbeto committed:
perf(ivy): chain multiple attribute instructions

Similarly to what we did in #31078, these changes implement chaining for the `attribute` instruction.

This PR resolves FW-1390.

kara commented:
[New Ivy presubmit](https://test.corp.google.com/ui#id=OCL:254763918:BASE:254793202:1561399343210:7ada801e)
